core/src/test/scala/unit/kafka/utils/TestUtils.scala

```diff
@@ -349,10 +349,13 @@ object TestUtils extends Logging {
       !hasSessionExpirationException},
       s"Can't create topic $topic")
 
+    // wait until we've got the expected partition size
+    waitUntilMetadataIsPropagatedWithExpectedSize(servers, topic, numPartitions)
```

Comment:
As mentioned in the other comment, this is probably good enough. The extra validations below seem like overkill.